File: benchmark/2026-09-17-7677581(latest)/micro/src/go/mb07.go
// MB-07 - Sorting: bottom-up iterative merge sort, ascending, stable, ping-pong buffers.
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"time"
)

// lcg is the Lehmer / MINSTD generator, frozen for every language in the suite.
type lcg struct {
	state int64
}

func (g *lcg) nextInt() int64 {
	g.state = (48271 * g.state) % 2147483647
	return g.state
}

// msort sorts a in place using buf as the alternate ping-pong buffer.
func msort(a, buf []int64) {
	n := len(a)
	src, dst := a, buf
	for width := 1; width < n; width *= 2 {
		for lo := 0; lo < n; lo += 2 * width {
			mid := min(lo+width, n)
			hi := min(lo+2*width, n)
			i, j, k := lo, mid, lo
			for i < mid && j < hi {
				if src[i] <= src[j] {
					dst[k] = src[i]
					i++
				} else {
					dst[k] = src[j]
					j++
				}
				k++
			}
			for i < mid {
				dst[k] = src[i]
				i++
				k++
			}
			for j < hi {
				dst[k] = src[j]
				j++
				k++
			}
		}
		src, dst = dst, src
	}
	if &src[0] != &a[0] {
		copy(a, src)
	}
}

type result struct {
	total, ssum, inv int64
}

// workload is the entire measured body: it re-seeds the generator and
// regenerates src, as methodology 06 section 5.2 requires of every steady
// iteration.
func workload() result {
	const (
		n = 2000000
		r = 4
	)

	gen := lcg{state: 20267917}
	src := make([]int64, n)
	for i := 0; i < n; i++ {
		src[i] = gen.nextInt()
	}
	buf := make([]int64, n)

	var total, ssum, inv int64
	for round := 0; round < r; round++ {
		src[round] = src[round] + 1 // anti-elimination
		a := slices.Clone(src)
		msort(a, buf)

		var chk int64
		for i := 0; i < n; i++ {
			chk = (chk*31 + a[i]%1000003) % 1000003
		}
		total = (total*7 + chk) % 1000003

		ssum = 0
		for i := 0; i < n; i++ {
			ssum = ssum + a[i]
		}
		for i := 1; i < n; i++ {
			if a[i-1] > a[i] {
				inv++
			}
		}
	}

	return result{total: total, ssum: ssum, inv: inv}
}

// parseMode reads the section 5.2 program mode from argv[1] and the steady
// iteration count K from argv[2] (default 7).
func parseMode() (string, int) {
	mode := "once"
	if len(os.Args) > 1 {
		mode = os.Args[1]
	}
	k := 7
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil && v > 0 {
			k = v
		}
	}
	return mode, k
}

func main() {
	mode, k := parseMode()
	var r result
	if mode == "steady" {
		for i := 0; i < k; i++ {
			t0 := time.Now() // section 5.2 pins time.Now/time.Since for Go
			r = workload()
			elapsed := time.Since(t0).Nanoseconds()
			// os.Stdout is unbuffered in Go, so this Printf is an immediate write.
			fmt.Printf("ITER %d %d\n", i, elapsed)
		}
	} else {
		r = workload()
	}
	fmt.Printf("MB07 %d %d %d\n", r.total, r.ssum, r.inv)
}

File: benchmark/2026-09-17-7677581(latest)/micro/src/swift/mb07.swift
// MB-07 - Sorting: bottom-up iterative merge sort, ascending, stable, ping-pong buffers.
import Darwin  // fflush(stdout) for the section 5.2 ITER lines

/// Lehmer / MINSTD generator, frozen for every language in the suite.
struct Lcg {
    private var state: Int

    init(seed: Int) {
        state = seed
    }

    mutating func nextInt() -> Int {
        state = (48271 * state) % 2147483647
        return state
    }
}

/// One bottom-up merge pass of the pinned algorithm: merge runs of `width`
/// from `src` into `dst`.
func mergePass(_ src: inout [Int], _ dst: inout [Int], _ n: Int, _ width: Int) {
    var lo = 0
    while lo < n {
        let mid = min(lo + width, n)
        let hi = min(lo + 2 * width, n)
        var i = lo
        var j = mid
        var k = lo
        while i < mid && j < hi {
            if src[i] <= src[j] {
                dst[k] = src[i]
                i += 1
            } else {
                dst[k] = src[j]
                j += 1
            }
            k += 1
        }
        while i < mid {
            dst[k] = src[i]
            i += 1
            k += 1
        }
        while j < hi {
            dst[k] = src[j]
            j += 1
            k += 1
        }
        lo += 2 * width
    }
}

/// Bottom-up merge sort of `a[0..<n]`; `buf` is the ping-pong partner buffer.
/// The source/destination roles alternate, exactly as the pinned pseudocode
/// specifies, and the mandated final copy-back runs when an odd number of
/// passes has left the sorted data in `buf` (21 passes at n = 2000000).
func msort(_ a: inout [Int], _ buf: inout [Int], _ n: Int) {
    var flipped = false
    var width = 1
    while width < n {
        if flipped {
            mergePass(&buf, &a, n, width)
        } else {
            mergePass(&a, &buf, n, width)
        }
        flipped = !flipped
        width *= 2
    }
    if flipped {
        // "if src is not a: copy src[0..n-1] into a[0..n-1]"
        a.replaceSubrange(0..<n, with: buf[0..<n])
    }
}

/// The whole workload body, generation included (methodology 06 section 4, MB-07).
/// Section 5.2 requires each steady iteration to re-seed and regenerate its input.
func workload() -> String {
    let n = 2000000
    let r = 4

    var gen = Lcg(seed: 20267917)
    var src = [Int](repeating: 0, count: n)
    for i in 0..<n {
        src[i] = gen.nextInt()
    }
    var buf = [Int](repeating: 0, count: n)

    var total = 0
    var ssum = 0
    var inv = 0
    for round in 0..<r {
        src[round] = src[round] + 1  // anti-elimination, part of the algorithm
        var a = src                  // independent copy of the unsorted input
        msort(&a, &buf, n)

        var chk = 0
        for i in 0..<n {
            chk = (chk * 31 + (a[i] % 1000003)) % 1000003
        }
        total = (total * 7 + chk) % 1000003

        ssum = 0
        for i in 0..<n {
            ssum = ssum + a[i]
        }
        for i in 1..<n where a[i - 1] > a[i] {
            inv += 1
        }
    }

    return "MB07 \(total) \(ssum) \(inv)"
}

// Section 5.2 program modes: argv[1] in {once, steady}, argv[2] = K (default 7).
func main() {
    let args = CommandLine.arguments
    let mode = args.count > 1 ? args[1] : "once"
    var line: String
    if mode == "steady" {
        var iterations = args.count > 2 ? (Int(args[2]) ?? 7) : 7
        if iterations < 1 {
            iterations = 7
        }
        let clock = ContinuousClock()
        line = ""
        for k in 0..<iterations {
            let elapsed = clock.measure {
                line = workload()
            }
            let parts = elapsed.components
            let ns = parts.seconds * 1_000_000_000 + parts.attoseconds / 1_000_000_000
            print("ITER \(k) \(ns)")
            fflush(stdout)  // section 5.2: each ITER line is flushed immediately
        }
    } else {
        line = workload()
    }
    print(line)
}

main()
